### PR TITLE
Add WebSocket wire protocol documentation

### PR DESCRIPTION
**Pull request by davidbrochart**
cc @jasongrout

### CODE REVIEW

File: docs/source/developers/websocket-protocols.rst [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+     - content
+     - buffer_0
+     - buffer_1
+     - ...
+
+See also the `Jupyter Client documentation <https://jupyter-client.readthedocs.io/en/stable/messaging.html#the-wire-protocol>`_.
+
+Note that a set of ZeroMQ sockets, one for each channel (``IOPub``, ``shell``, etc.), are multiplexed into one WebSocket. Thus, the channel name must be encoded in WebSocket messages.
+
+WebSocket protocol negociation
```
Comment (blink1073) [inline docs/source/developers/websocket-protocols.rst:37]:
```suggestion
WebSocket protocol negotiation
```

File: docs/source/developers/websocket-protocols.rst [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+     - content
+     - buffer_0
+     - buffer_1
+     - ...
+
+Where:
+
+* ``offset_number`` is a 64-bit (little endian) unsigned integer.
+* ``offset_0`` to ``offset_n`` are 64-bit (little endian) unsigned integers (with ``n=offset_number-1``).
+* The message can be deserialized from the ``bin_msg`` serialized message as follows (Python code):
```
Comment (jasongrout) [inline docs/source/developers/websocket-protocols.rst:139]:
```suggestion
* ``channel`` is a UTF-8 encoded string containing the channel for the message (shell, iopub, etc.)
* ``header``, ``parent_header``, ``metadata``, and ``content`` are UTF-8 encoded JSON text representing the given part of a message in the Jupyter message protocol.
* ``offset_n`` is the number of bytes in the message.
* The message can be deserialized from the ``bin_msg`` serialized message as follows (Python code):
```

File: docs/source/developers/websocket-protocols.rst [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+.. code-block:: python
+
+    channel = bin_msg[offset_0:offset_1]
+    header = bin_msg[offset_1:offset_2]
+    parent_header = bin_msg[offset_2:offset_3]
+    metadata = bin_msg[offset_3:offset_4]
+    content = bin_msg[offset_4:offset_5]
```
Comment (jasongrout) [inline docs/source/developers/websocket-protocols.rst:147]:
```suggestion
    import json
    channel = bin_msg[offset_0:offset_1].decode('utf-8')
    header = json.loads(bin_msg[offset_1:offset_2])
    parent_header = json.loads(bin_msg[offset_2:offset_3])
    metadata = json.loads(bin_msg[offset_3:offset_4])
    content = json.loads(bin_msg[offset_4:offset_5])
```

File: docs/source/developers/websocket-protocols.rst [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    buffer_0 = bin_msg[offset_5:offset_6]
+    buffer_1 = bin_msg[offset_6:offset_7]
+    # ...
```
Comment (jasongrout) [inline docs/source/developers/websocket-protocols.rst:150]:
```suggestion
    # ...
    last_buffer = bin_msg[offset_(n-1):offset_n]
```

**Review by jasongrout — APPROVED** [review]
Thanks! I added a few suggestions for clarification.

**Comment by davidbrochart** [comment]
Thanks @jasongrout !

**Pull request merged**